File: draft/tools/season_completeness.js
/* TERRITORY: A. ONE ANSWER TO "HAS THIS SEASON ACTUALLY BEEN PLAYED?"
 *
 * ── WHY THIS IS A MODULE AND NOT THREE COPIES ──────────────────────────────
 *
 * Register 419: `objective_dp.js` guarded its season loop with
 * `if (!season.weeks || ...) return;`. That held while 2026 carried no weeks.
 * After the 08-22 draft the 2026 scaffolding landed EIGHTEEN WEEKS OF ZEROS —
 * week 1 is 09-10, nothing has been played — so 2026 walked through the guard
 * and added ten seats grading 0 for every owner and every arm. Every mean the
 * tool reported was diluted by exactly 30/40: +45.84/+29.33 read as
 * +34.38/+22.00, 0.75x to the decimal. Its control A refused to report, which
 * is the only reason anyone found out.
 *
 * `ungraded_season_leak.py` then measured the rest of the registry and found
 * the same contamination in `roster_builder_replay` (including a PREDICTION
 * LEDGER grade, P215's mean_delta −15.3 where the truth is −20.41) and in
 * `waiver_realized_level`. Three tools, one predicate.
 *
 * Register 408 is the reason this is a require and not a fourth paste: the
 * duplicate-arm guard was shipped four times by hand THAT SAME WEEK, by the
 * person quoting the lesson, and had to be extracted within the hour.
 *
 * ── THE PREDICATE IS "COMPLETE", AND THE DIFFERENCE IS A DATE ──────────────
 *
 * ⚠️ The obvious version asks whether ANY week has realized points. That is
 * correct today and becomes WRONG ON 2026-09-10, when week 1 is played: 2026
 * would satisfy it and enter a FULL-SEASON grade on 1 of 18 weeks — a subtler
 * error than the zeros, arriving on a schedule with nobody watching. A season
 * qualifies only when EVERY week it carries has been played.
 *
 * Measured 2026-08-29: 2023, 2024 and 2025 each carry 18 of 18 scored weeks;
 * 2026 carries 0 of 18.
 *
 * ⛔ THIS IS FOR SEASON-LEVEL GRADES. A weekly instrument that legitimately
 * wants partial data should NOT use it — and an artifact that is ABOUT the
 * current season (opponent_need_2026 requires 2026 by construction) must not
 * either. Filtering is not always the right answer; knowing is.
 */
'use strict';

/* A season counts only if every week it carries has somebody scoring in it. */
function isCompleteSeason(season) {
  const weeks = Object.values((season && season.weeks) || {});
  if (!weeks.length) return false;
  return weeks.every(entries => (entries || []).some(
    e => Number(e && e.points) > 0));
}

/* Partition, so the caller can REPORT the exclusion instead of silently
 * dropping it. An exclusion nobody can see is the defect 419 was built out
 * of, so the shape of this function makes announcing it the easy path. */
function splitByCompleteness(seasons) {
  const complete = [], incomplete = [];
  (seasons || []).forEach(s => (isCompleteSeason(s) ? complete : incomplete).push(s));
  return { complete, incomplete };
}

function selfTest() {
  let pass = 0, fail = 0;
  const ck = (n, ok, d) => { ok ? (pass++, console.log('PASS  ' + n))
    : (fail++, console.log('FAIL  ' + n + (d !== undefined ? '\n        ' + JSON.stringify(d).slice(0, 200) : ''))); };

  const scored = n => { const w = {}; for (let i = 1; i <= n; i++) w[i] = [{ points: 101.5 }]; return { weeks: w }; };
  const zeros = n => { const w = {}; for (let i = 1; i <= n; i++) w[i] = [{ points: 0 }]; return { weeks: w }; };

  ck('KNOWN NEGATIVE — 18 weeks of ZEROS is not a played season (the 2026 shape)',
    isCompleteSeason(zeros(18)) === false);
  ck('KNOWN POSITIVE — 18 weeks all scored IS (the 2023-25 shape)',
    isCompleteSeason(scored(18)) === true);

  /* THE DATE TRAP, as a test rather than a comment. */
  const oneOf18 = zeros(18);
  oneOf18.weeks[1] = [{ points: 101.5 }];
  ck('⭐ ONE scored week of 18 is still INCOMPLETE — this is the assertion that '
    + 'stops 09-10 from silently re-opening register 419',
    isCompleteSeason(oneOf18) === false);

  ck('no weeks at all is incomplete', isCompleteSeason({ weeks: {} }) === false);
  ck('a missing weeks key does not throw', isCompleteSeason({}) === false);
  ck('null does not throw', isCompleteSeason(null) === false);
  ck('a week whose entries are null does not throw and is incomplete',
    isCompleteSeason({ weeks: { 1: null } }) === false);

  const { complete, incomplete } = splitByCompleteness([scored(18), zeros(18), scored(18)]);
  ck('splitByCompleteness partitions and loses nothing',
    complete.length === 2 && incomplete.length === 1);

  /* Against the REAL file, so the module cannot drift from the data it is
   * about (register 121: a control written against invented input). */
  try {
    const path = require('path');
    const H = JSON.parse(require('fs').readFileSync(
      path.join(__dirname, '..', 'data', 'league_history.json'), 'utf8'));
    const s = {};
    H.seasons.forEach(x => { s[x.season] = isCompleteSeason(x); });
    ck('AGAINST THE LIVE FILE — 2023/2024/2025 complete, 2026 not',
      s['2023'] === true && s['2024'] === true && s['2025'] === true && s['2026'] === false, s);
  } catch (e) {
    ck('AGAINST THE LIVE FILE — league_history.json readable', false, String(e.message));
  }

  console.log('\n' + pass + '/' + (pass + fail) + ' self-tests passed');
  return fail ? 1 : 0;
}

if (require.main === module) process.exit(selfTest());
module.exports = { isCompleteSeason, splitByCompleteness };

File: draft/tools/season_completeness.py
# TERRITORY: A
"""ONE ANSWER TO "HAS THIS SEASON ACTUALLY BEEN PLAYED?" — the Python side.

`draft/tools/season_completeness.js` is the JavaScript side. Two languages
need the predicate, so there are two implementations and that is a risk, not
a design: `draft/tests/test_season_completeness_agrees.py` asserts they give
the SAME answer for every season in the live `league_history.json`, so a
divergence is a red test rather than two tools quietly disagreeing about
which seasons count.

── WHY IT EXISTS (register 419) ────────────────────────────────────────────

`objective_dp.js` guarded its season loop with `if (!season.weeks || ...)`.
That held while 2026 carried no weeks. After the 08-22 draft the scaffolding
landed EIGHTEEN WEEKS OF ZEROS — week 1 is 09-10 — and 2026 also has a
150-pick draft, so it walked through and added ten seats grading 0 for every
owner and every arm, diluting every mean by exactly 30/40.

Register 420 then found the same shape in `exp25_deadzone.load_picks()`,
which is shared by three studies: it loops `for s in hist["seasons"]` with no
guard and attaches `realized` points per season, so every 2026 pick enters
carrying a realized value of 0.0.

── THE PREDICATE IS "COMPLETE", AND THE DIFFERENCE IS A DATE ───────────────

⚠️ The obvious version asks whether ANY week has realized points. That is
correct today and becomes WRONG ON 2026-09-10, when week 1 is played: 2026
would satisfy it and enter a full-season measurement on 1 of 18 weeks — a
subtler error than the zeros, arriving on a schedule with nobody watching. A
season qualifies only when EVERY week it carries has been played.

Measured 2026-08-29: 2023, 2024 and 2025 carry 18 of 18 scored weeks; 2026
carries 0 of 18.

⛔ THIS IS FOR MEASUREMENTS THAT NEED REALIZED OUTCOMES. A study that reads
only DRAFT PICKS should NOT use it — the 2026 draft really happened on 08-22
and is legitimate evidence of how owners draft. `opponent_profiles.py` is
exactly that case and is deliberately left alone. Filtering is not always the
right answer; knowing is.

Run: python3 draft/tools/season_completeness.py   (self-test)
"""
from __future__ import annotations

from typing import Any


def is_complete_season(season: Any) -> bool:
    """Every week the season carries has somebody scoring in it."""
    weeks = list(((season or {}).get("weeks") or {}).values())
    if not weeks:
        return False
    return all(
        any(float((e or {}).get("points") or 0) > 0 for e in (entries or []))
        for entries in weeks
    )


def split_by_completeness(seasons) -> tuple[list, list]:
    """(complete, incomplete). Returns both so the caller can ANNOUNCE the
    exclusion rather than drop it silently — an exclusion nobody can see is
    the defect register 419 was built out of."""
    complete, incomplete = [], []
    for s in seasons or []:
        (complete if is_complete_season(s) else incomplete).append(s)
    return complete, incomplete


def _self_test() -> int:
    import json
    import pathlib

    passed = failed = 0

    def ck(name, ok, detail=None):
        nonlocal passed, failed
        if ok:
            passed += 1
            print(f"PASS  {name}")
        else:
            failed += 1
            print(f"FAIL  {name}" + (f"\n        {detail!r}"[:240] if detail is not None else ""))

    def scored(n):
        return {"weeks": {str(i): [{"points": 101.5}] for i in range(1, n + 1)}}

    def zeros(n):
        return {"weeks": {str(i): [{"points": 0}] for i in range(1, n + 1)}}

    ck("KNOWN NEGATIVE — 18 weeks of ZEROS is not a played season (the 2026 shape)",
       is_complete_season(zeros(18)) is False)
    ck("KNOWN POSITIVE — 18 weeks all scored IS (the 2023-25 shape)",
       is_complete_season(scored(18)) is True)

    one = zeros(18)
    one["weeks"]["1"] = [{"points": 101.5}]
    ck("⭐ ONE scored week of 18 is still INCOMPLETE — the assertion that stops "
       "09-10 from silently re-opening register 419", is_complete_season(one) is False)

    ck("no weeks at all is incomplete", is_complete_season({"weeks": {}}) is False)
    ck("a missing weeks key does not raise", is_complete_season({}) is False)
    ck("None does not raise", is_complete_season(None) is False)
    ck("a week whose entries are None does not raise and is incomplete",
       is_complete_season({"weeks": {"1": None}}) is False)

    c, i = split_by_completeness([scored(18), zeros(18), scored(18)])
    ck("split_by_completeness partitions and loses nothing", len(c) == 2 and len(i) == 1)

    root = pathlib.Path(__file__).resolve().parent.parent.parent
    try:
        hist = json.loads((root / "draft" / "data" / "league_history.json").read_text())
        got = {s["season"]: is_complete_season(s) for s in hist["seasons"]}
        ck("AGAINST THE LIVE FILE — 2023/2024/2025 complete, 2026 not",
           got.get("2023") and got.get("2024") and got.get("2025")
           and got.get("2026") is False, got)
    except OSError as e:
        ck("AGAINST THE LIVE FILE — league_history.json readable", False, str(e))

    print(f"\n{passed}/{passed + failed} self-tests passed")
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(_self_test())
